File: app/src/main/java/io/kuthorx/github/home_sport/WorkoutEngine.java
package io.kuthorx.github.home_sport;

import java.util.List;

public final class WorkoutEngine {
    private final List<WorkoutStep> steps;
    private int stepIndex;
    private int secondsRemaining;
    private boolean paused;
    private boolean complete;

    public WorkoutEngine(WorkoutPlan plan) {
        steps = plan.getSteps();
        secondsRemaining = steps.get(0).getDurationSeconds();
    }

    private WorkoutEngine(
            WorkoutPlan plan,
            int stepIndex,
            int secondsRemaining,
            boolean paused,
            boolean complete
    ) {
        steps = plan.getSteps();
        if (complete) {
            if (stepIndex != steps.size() || secondsRemaining != 0 || paused) {
                throw new IllegalArgumentException("invalid completed workout state");
            }
        } else if (stepIndex < 0 || stepIndex >= steps.size()
                || secondsRemaining <= 0
                || secondsRemaining > steps.get(stepIndex).getDurationSeconds()) {
            throw new IllegalArgumentException("invalid active workout state");
        }
        this.stepIndex = stepIndex;
        this.secondsRemaining = secondsRemaining;
        this.paused = paused;
        this.complete = complete;
    }

    public static WorkoutEngine restore(
            WorkoutPlan plan,
            int stepIndex,
            int secondsRemaining,
            boolean paused,
            boolean complete
    ) {
        return new WorkoutEngine(plan, stepIndex, secondsRemaining, paused, complete);
    }

    public void tick() {
        if (paused || complete) {
            return;
        }
        secondsRemaining--;
        if (secondsRemaining == 0) {
            advance();
        }
    }

    public void pause() {
        if (!complete) {
            paused = true;
        }
    }

    public void resume() {
        paused = false;
    }

    public void skip() {
        if (!complete) {
            advance();
        }
    }

    private void advance() {
        stepIndex++;
        if (stepIndex == steps.size()) {
            complete = true;
            paused = false;
            secondsRemaining = 0;
            return;
        }
        secondsRemaining = currentStep().getDurationSeconds();
    }

    public WorkoutStep currentStep() {
        return complete ? null : steps.get(stepIndex);
    }

    public WorkoutStep.Phase getPhase() {
        return complete ? WorkoutStep.Phase.COMPLETED : currentStep().getPhase();
    }

    public int getSecondsRemaining() { return secondsRemaining; }
    public int getStepIndex() { return stepIndex; }
    public int getCurrentSet() { return complete ? 0 : currentStep().getSetNumber(); }
    public int getTotalSets() { return complete ? 0 : currentStep().getTotalSets(); }
    public int getCurrentRest() { return complete ? getTotalRests() : currentStep().getRestNumber(); }
    public int getTotalRests() { return steps.get(0).getTotalRests(); }
    public boolean isPaused() { return paused; }
    public boolean isComplete() { return complete; }

}

File: app/src/main/java/io/kuthorx/github/home_sport/WorkoutEngine.kt
package io.kuthorx.github.home_sport

class WorkoutEngine private constructor(
    plan: WorkoutPlan,
    private var mutableStepIndex: Int,
    private var mutableSecondsRemaining: Int,
    private var mutablePaused: Boolean,
    private var mutableComplete: Boolean,
    private var mutableSkippedSteps: Int,
) {
    private val steps = plan.steps

    val stepIndex: Int
        get() = mutableStepIndex

    val secondsRemaining: Int
        get() = mutableSecondsRemaining

    val isPaused: Boolean
        get() = mutablePaused

    val isComplete: Boolean
        get() = mutableComplete

    val skippedSteps: Int
        get() = mutableSkippedSteps

    constructor(plan: WorkoutPlan) : this(
        plan,
        mutableStepIndex = 0,
        mutableSecondsRemaining = plan.steps[0].durationSeconds,
        mutablePaused = false,
        mutableComplete = false,
        mutableSkippedSteps = 0,
    )

    init {
        require(mutableSkippedSteps >= 0) { "skippedSteps must not be negative" }
        if (mutableComplete) {
            require(mutableStepIndex == steps.size && mutableSecondsRemaining == 0 && !mutablePaused) {
                "invalid completed workout state"
            }
        } else {
            require(
                mutableStepIndex in steps.indices &&
                    mutableSecondsRemaining > 0 &&
                    mutableSecondsRemaining <= steps[mutableStepIndex].durationSeconds,
            ) { "invalid active workout state" }
        }
    }

    fun tick() {
        if (mutablePaused || mutableComplete) return
        mutableSecondsRemaining--
        if (mutableSecondsRemaining == 0) advance()
    }

    fun pause() {
        if (!mutableComplete) mutablePaused = true
    }

    fun resume() {
        mutablePaused = false
    }

    fun skip() {
        if (!mutableComplete) {
            mutableSkippedSteps++
            advance()
        }
    }

    private fun advance() {
        mutableStepIndex++
        if (mutableStepIndex == steps.size) {
            mutableComplete = true
            mutablePaused = false
            mutableSecondsRemaining = 0
            return
        }
        mutableSecondsRemaining = currentStep()!!.durationSeconds
    }

    fun currentStep(): WorkoutStep? = if (mutableComplete) null else steps[mutableStepIndex]

    val phase: WorkoutStep.Phase
        get() = if (mutableComplete) WorkoutStep.Phase.COMPLETED else currentStep()!!.phase

    val currentSet: Int
        get() = if (mutableComplete) 0 else currentStep()!!.setNumber

    val totalSets: Int
        get() = if (mutableComplete) 0 else currentStep()!!.totalSets

    val currentRest: Int
        get() = if (mutableComplete) totalRests else currentStep()!!.restNumber

    val totalRests: Int
        get() = steps[0].totalRests

    companion object {
        @JvmStatic
        fun restore(
            plan: WorkoutPlan,
            stepIndex: Int,
            secondsRemaining: Int,
            paused: Boolean,
            complete: Boolean,
        ): WorkoutEngine = restore(plan, stepIndex, secondsRemaining, paused, complete, 0)

        @JvmStatic
        fun restore(
            plan: WorkoutPlan,
            stepIndex: Int,
            secondsRemaining: Int,
            paused: Boolean,
            complete: Boolean,
            skippedSteps: Int,
        ): WorkoutEngine = WorkoutEngine(
            plan,
            stepIndex,
            secondsRemaining,
            paused,
            complete,
            skippedSteps,
        )
    }
}
